### ISSUE TITLE
Wrap AbstractArray rather than Array in SampleBuf?

### ISSUE THREAD

**Issue by mchitre** [JuliaAudio/SampledSignals.jl#60]
Currently `SampleBuf` wraps an `Array`. This disallows wrapping of views or other array-like data types, for example. Is there any strong reason why not wrap an `AbstractArray` instead?

**Comment by mchitre** [comment]
Fixing this might allow nice adding constructs like `@view buf[pos1:pos2]` for `buf::SampleBuf` to also be a `SampleBuf`, and therefore retain sampling rate metadata.

**Comment by mchitre** [comment]
I've implemented this for testing, and seems to work well:
https://github.com/mchitre/SampledSignals.jl/commit/69078ab6fac4431098015aa2ea5fe9189f9a964c

If you don't see any problems with this change, and are open to it, I can raise a PR.

**Comment by ssfrr** [comment]
Yeah, I don't have any problem with that at all, seems like a good improvement.

For your `view` example you'd have a `SampleBuf` that wrapped a view into the original data, right?

**Comment by mchitre** [comment]
Correct

**Comment by ssfrr** [comment]
Great, that's quite nice. Thanks!

**Comment by ssfrr** [comment]
Also note that changing this `view` behavior would be a breaking change, so I'll add it to the list for the next breaking release.

**Comment by mchitre** [comment]
I've included the `AbstractArray` change in the same PR, as it is quite minor. I'll do a separate PR later for the view change, once I've implemented, tested and am happy with it being robust.

**Comment by haberdashPI** [comment]
I think I did something similar with the large PR I wrote a while back, might be worth checking in there to see if there's anything else you want to use.

**Comment by mchitre** [comment]
Actually turned out to be simple enough. I've added it to the same PR but in a different commit (in case you want to release them in separate version). I've added a test case to cover it too.

**Comment by mchitre** [comment]
I realize that wrapping an `AbstractArray` may not be such a good idea for type inference [[1]](https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-abstract-container-1). I haven't thought through this fully, and figured it was better to drop this from the PR until the best design approach discussed/explored.

**Comment by ssfrr** [comment]
Yeah, rather that actually wrapping AbstractArray the wrapped type should be given as a type parameter. Otherwise indexing etc. will be very slow.

**Comment by mchitre** [comment]
Yup. But since this is encoded in `AbstractSampleBuf{T,N}`, the implications with the remaining code base may be substantial.

**Comment by pjabardo** [comment]
I just created a pull request that allows the use of `AbstractArray`s as data field in `SampleBuf` and `SpectrumBuf`. The tests are passing and I don't believe that this will result in performance issues. The approach used here is the same as as the one used in the package `AxisArrays`.

I believe that the required Julia version is now 1.0 since triangular dispatch is used. But further testing is necessary.

I did some preliminary testing using `StaticArrays` and `FixedSizeArrays` and they appear to be supported. 

`view`works but it would be nice when using `view` with an `OrdinalRange` the code could return a `SampleBuf` with a corrected samplerate.
